# Phase 1 project

This note book is designed to analyse data National Transportation Safety Board that includes aviation accident data from 1962 to 2023 about civil aviation accidents and selected incidents in the United States and international waters  and identify the lowest-risk aircraft for the company to purchase and operate in the commercial and private aviation sectors, providing actionable insights to guide decision-making for the new aviation division

Specifically, this will cover:

* Using pandas to filter data
* Using pandas to handle missing values
* Using matplotlib to create a graph using data from a dataframe

##  Task: Analyze Airplane Data

### The following questions are to be answered from the analysis of the data:
1. Determine which aircraft are the lowest risk for the company to start this new business endeavor.
2. Identify actionable insights to help the head of the new aviation division decide which aircraft to purchase.

### Data Understanding

In this repository under the file path `AviationData.csv` there is a CSV file containing information about events that occured on various aircrafts

There is also `USState_Codes.csv` which contains short form identifies of various states in the United states




### Requirements

#### 1. Filter Data to Relevant Columns

#### 2. Filter Data to Relevant Rows

#### 3. Drop Rows with Missing Values

#### 4. Identify lowest risk plane

#### 5. Plot charts to support the assement above

### Setup

In the cell below we import the relevant libraries



In [57]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

## Open up the CSV file as a dataframe called `df ` 

In the cell below we open the `AviationData.csv` file as a df and print the first 5 rows to get an overview of the data. The data is encoded in latin and the standard `pd.read_csv('data')`command doesn't work hence the need to add
`encoding=latin-1` arguement and setting `low_memory=False` to stop the printing of warnings

In [70]:
df=pd.read_csv('AviationData.csv',encoding='latin-1',low_memory=False)
df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

## Identify information about the data set
In the cells below we identify information about the dataset such number of columns and rows and datatypes represented by each column

In [37]:
#Get datatypes of each column and number of rows and columns
df.shape

(88889, 31)

In [40]:
#Get datatypes of each column and number of rows and columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

## Overview analysis of the dataset 
* From the cells above, we can see that the dataset has `88889` rows and `31` columns
* The data has various datatypes
* Some columns seems to have null values

#### In the cell below, we will try and identify the number of null values in the dataset

In [71]:
#Percentage of null values from each column
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.058500
Country                    0.254250
Latitude                  61.320298
Longitude                 61.330423
Airport.Code              43.469946
Airport.Name              40.611324
Injury.Severity            1.124999
Aircraft.damage            3.593246
Aircraft.Category         63.677170
Registration.Number        1.481623
Make                       0.070875
Model                      0.103500
Amateur.Built              0.114750
Number.of.Engines          6.844491
Engine.Type                7.961615
FAR.Description           63.974170
Schedule                  85.845268
Purpose.of.flight          6.965991
Air.carrier               81.271023
Total.Fatal.Injuries      12.826109
Total.Serious.Injuries    14.073732
Total.Minor.Injuries      13.424608
Total.Uninjured            6.650992
Weather.Condition          5

## Analysis of the null values in the dataset
* From the cell above we can see almost all the cells have null values apart from 4 columns are `Event.Id`,`Investigation.Type`,`Accident.Number` and `Event.Date`
* We will drop all columns with null percentage value of over 40 % from the data set.This columns include  `Latitude`,`Longitude`,`Airport.Code`,`Airport.Name`,`Aircraft.Category`,`FAR.Description`,`Schedule`,`Air.carrier`.

We want to drop the above columns because there more or close to more missing data than available data, which makes it difficult to extract meaningful insights. Retaining such columns may introduce bias or inaccuracies in our analysis.

In [72]:
# List of columns to drop
columns_to_drop = [
    'Latitude', 'Longitude', 'Airport.Code', 'Airport.Name', 
    'Aircraft.Category', 'FAR.Description', 'Schedule', 'Air.carrier'
]

# Drop the columns from the DataFrame
df = df.drop(columns=columns_to_drop)

In [60]:
#Verifying the result that all columns with null percentage value of over 40% are dropped
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.058500
Country                    0.254250
Injury.Severity            1.124999
Aircraft.damage            3.593246
Registration.Number        1.481623
Make                       0.070875
Model                      0.103500
Amateur.Built              0.114750
Number.of.Engines          6.844491
Engine.Type                7.961615
Purpose.of.flight          6.965991
Total.Fatal.Injuries      12.826109
Total.Serious.Injuries    14.073732
Total.Minor.Injuries      13.424608
Total.Uninjured            6.650992
Weather.Condition          5.053494
Broad.phase.of.flight     30.560587
Report.Status              7.178616
Publication.Date          15.492356
dtype: float64

# Dealing with remaining null values

## Location column

First we will start with columns that have non-numerical data such as `Location` 
We will drop all rows where column location is null. This is because the column `Location`  has a null percentage of  0.058500%. Dropping these rows means we are only losing a small portion of ourdata, which is often acceptable.

In [73]:
# Drop rows where 'Location' is null
df.dropna(subset=['Location'], inplace=True)

In [75]:
# Verify the result
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.000000
Country                    0.253273
Injury.Severity            1.125657
Aircraft.damage            3.587469
Registration.Number        1.457726
Make                       0.070916
Model                      0.103560
Amateur.Built              0.110314
Number.of.Engines          6.815854
Engine.Type                7.959521
Purpose.of.flight          6.958812
Total.Fatal.Injuries      12.817857
Total.Serious.Injuries    14.061709
Total.Minor.Injuries      13.411079
Total.Uninjured            6.642503
Weather.Condition          5.050823
Broad.phase.of.flight     30.535700
Report.Status              7.182818
Publication.Date          15.487916
dtype: float64

## Country column

Next we will move to the `Country` column. 

Similar to the Location clolumn, We will drop all rows where column `country` is null. This is because the column `Location`  has a null percentage of  0.253273%. Dropping these rows means we are only losing a small portion of ourdata, which is often acceptable.

In [77]:
# Drop rows where 'Country' is null
df.dropna(subset=['Country'], inplace=True)

In [78]:
# Verify the result
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.000000
Country                    0.000000
Injury.Severity            1.128515
Aircraft.damage            3.578522
Registration.Number        1.455785
Make                       0.069968
Model                      0.102695
Amateur.Built              0.110595
Number.of.Engines          6.825261
Engine.Type                7.972961
Purpose.of.flight          6.968582
Total.Fatal.Injuries      12.837990
Total.Serious.Injuries    14.086128
Total.Minor.Injuries      13.433846
Total.Uninjured            6.649212
Weather.Condition          5.058006
Broad.phase.of.flight     30.601950
Report.Status              7.201056
Publication.Date          15.462917
dtype: float64

## Injury.Severity column

Next we will move to the `Injury.Severity` column. 

Let's look at the type of data we are dealing with:

In [79]:
injury_severity_percentage = df['Injury.Severity'].value_counts(normalize=True) * 100
injury_severity_percentage

Non-Fatal     76.696115
Fatal(1)       7.004748
Fatal          6.006027
Fatal(2)       4.210610
Incident       2.496233
                ...    
Fatal(156)     0.001141
Fatal(141)     0.001141
Fatal(80)      0.001141
Fatal(57)      0.001141
Fatal(49)      0.001141
Name: Injury.Severity, Length: 109, dtype: float64

Seems like the data should have two categories which are `fatal` and `non-fatal`. But the way data has been stored the rows categorised as `fatal` have an integer placed in brackets making it seem like there more than two categories yet they referring to the same thing. We will clean up the rows by replacing all rows categorised as `fatal(int)` to `fatal`

In [94]:
# Replace 'Fatal(number)' with 'Fatal'
df['Injury.Severity'] = df['Injury.Severity'].str.replace(r'Fatal\(\d+\)', 'Fatal', regex=True)

# Verify the replacement
df['Injury.Severity'].value_counts(normalize=True) * 100

Non-Fatal      77.236126
Fatal          20.628399
Incident        1.617498
Minor           0.240153
Serious         0.177760
Unavailable     0.100064
Name: Injury.Severity, dtype: float64

### Analysis of `Injury.Severity` column
From the above verification we see that most injuries are Non-fatal then Followed by Fatal and we have unavailable which represents the null values. We will drop all rows that have null values since we are only losing a small portion of ourdata, which is often acceptable.

In [81]:
# Drop rows where 'Injury.Severity' is null
df.dropna(subset=['Injury.Severity'], inplace=True)

In [82]:
# Verify the result
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.000000
Country                    0.000000
Injury.Severity            0.000000
Aircraft.damage            3.042962
Registration.Number        1.446149
Make                       0.051363
Model                      0.081039
Amateur.Built              0.111857
Number.of.Engines          6.080217
Engine.Type                7.143999
Purpose.of.flight          5.964936
Total.Fatal.Injuries      12.984523
Total.Serious.Injuries    14.246907
Total.Minor.Injuries      13.587180
Total.Uninjured            6.725106
Weather.Condition          4.074784
Broad.phase.of.flight     29.809843
Report.Status              6.245720
Publication.Date          15.330092
dtype: float64

## Aircraft.damage column

Next we will move to the `Aircraft.damage` column. 

The column has 3% missing value percentage. Let's look at how the data is categorised in this column

In [91]:
aircraft_damage_percentage = df['Aircraft.damage'].value_counts(normalize=True) * 100
aircraft_damage_percentage

Substantial    75.148918
Destroyed      21.726744
Minor           3.016034
Unknown         0.108304
Name: Aircraft.damage, dtype: float64

### Analysis of `Aircraft.damage` column

The data seems to be categorical where we have three categories which are Substantial, Destroyed,Unknown and Minor 


In [85]:
# Drop rows where 'Aircraft.damage' is null
df.dropna(subset=['Aircraft.damage'], inplace=True)

In [87]:
# Verify the result
df.isna().sum()/ len(df) * 100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.000000
Country                    0.000000
Injury.Severity            0.000000
Aircraft.damage            0.000000
Registration.Number        1.299649
Make                       0.042380
Model                      0.072988
Amateur.Built              0.096532
Number.of.Engines          5.323382
Engine.Type                6.407600
Purpose.of.flight          4.823064
Total.Fatal.Injuries      12.826972
Total.Serious.Injuries    14.320863
Total.Minor.Injuries      13.486215
Total.Uninjured            6.826690
Weather.Condition          3.404516
Broad.phase.of.flight     29.305677
Report.Status              5.750712
Publication.Date          15.330916
dtype: float64

## Registration.Number column

Next we will move to the `Registration.Number` column. 

The column has 1.2% missing value percentage. Let's look at how the data is categorised in this column

In [96]:
registration_number_percentage = df['Registration.Number'].value_counts(normalize=True) * 100
registration_number_percentage

NONE      0.404332
UNREG     0.145512
None      0.077527
N20752    0.008349
N4101E    0.007156
            ...   
N365R     0.001193
N711MZ    0.001193
C-GPJH    0.001193
N214GP    0.001193
N22101    0.001193
Name: Registration.Number, Length: 75884, dtype: float64

### Analysis of `Registration.Number` column

The data seems to be unique such that each aircraft had a unique registration number and since almost all the rows had the registration number filled   
